### PR TITLE
irmin: use a more compact accumulator for unique tree traversals

### PR DESCRIPTION
**Pull request by CraigFe**
This PR improves the memory usage of ``Tree.fold ~unique:`True`` by a factor of two by introducing a more efficient hashset implementation for small fixed-length strings. This is a first step towards factoring the Tezos snapshot export logic into Irmin core (and so considerably simplifying `lib_context`). Provided in two commits:

1. introduces a new internal `irmin.data` library containing the hashset implementation. The hashset uses open addressing inside a bigstring with a relatively high maximum load factor in order to get close-to-optimal compactness without much runtime overhead.

1. switches `Tree.fold` to use this hashset implementation rather than the previous `(hash, unit) Stdlib.Hashtbl.t`.

The implementation comes with a basic benchmark that compares it to `Stdlib.Hashbtl.t`:

output

For now I've also included a Python script that takes the `.csv` output from the benchmark and generates the above graphs.

### CODE REVIEW

File: src/irmin/data/fixed_size_string_set.ml [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+           neither"
+          __MODULE__
+  in
+  let slot_count =
+    let rec aux n =
+      if n >= initial_capacity then n
+      else if n * 2 > Sys.max_array_length then n
+      else aux (n * 2)
+    in
+    aux 2
```
Comment (Ngoguey42) [inline src/irmin/data/fixed_size_string_set.ml:127]: I suggest that `let slot_count = initial_capacity` and that this function raises an `Invalid_arg` exception if `initial_capacity` is not a power of 2

File: src/irmin/data/fixed_size_string_set.mli [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+type t
+(** The type of mutable sets of fixed-length strings. *)
+
+type elt := string
+
+val create :
+  elt_length:int ->
+  ?initial_capacity:int ->
+  ?hash:(elt -> int) ->
+  ?hash_substring:(Bigstringaf.t -> off:int -> len:int -> int) ->
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/data/fixed_size_string_set.mli:26]: AFAIK, the compiler doesn't provide `int64` intrinsics for operating with bigarrays, so it's not possible to build bigstrings large enough to overflow an `int` anyway.

File: src/irmin/data/fixed_size_string_set.mli [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+      functions must be equivalent. Defaults to [Hashtbl.hash] (and an
+      equivalent function on substrings).
+
+    - [null]: a string of size [elt_length] that is guaranteed to never be added
+      to the hashset. Passing this string to {!add} or {!mem} after creating the
+      hashset will result in an exception being raised. *)
+
+include Hashset.S with type t := t and type elt := elt
+
+val invariant : (elt -> unit) -> t -> unit
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/data/fixed_size_string_set.mli:51]: Good idea, done:

```ocaml
val invariant : (elt -> unit) -> t -> unit
(** [invariant f t] checks the internal invariants of [t] and calls [f] on every
    element contained within. Exposed for testing. *)
```

This pattern was taken from various JS libs: https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/Invariant/index.html.

File: src/irmin/data/fixed_size_string_set.ml (hunk 1 above)
Comment (Ngoguey42) [reply src/irmin/data/fixed_size_string_set.ml:127]: I see.

I would rename `initial_capacity` to `initial_slots`, documenting in the docstring that `initial_slots` is not the initial capacity of the set.

File: src/irmin/data/fixed_size_string_set.mli (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ngoguey42) [reply src/irmin/data/fixed_size_string_set.mli:51]: Thanks :)

**Comment by icristescu** [comment]
@CraigFe can you move the python scripts (maybe at https://github.com/tarides/tezos-storage-bench or https://github.com/tarides/irmin-tezos?) so that we can merge this

**Comment by CraigFe** [comment]
Rebased and dealt with the outstanding comments. Will merge once the CI is passing to unblock https://github.com/mirage/irmin/pull/1757.

**Pull request merged**